src/blueapi/service/authentication.py

```diff
@@ -0,0 +1,231 @@
+from __future__ import annotations
+
+import base64
+import os
+import time
+from abc import ABC, abstractmethod
+from functools import cached_property
+from http import HTTPStatus
+from pathlib import Path
+from typing import Any, cast
+
+import jwt
+import requests
+from pydantic import TypeAdapter
+from requests.auth import AuthBase
+
+from blueapi.config import OIDCConfig
+from blueapi.service.model import Cache
+
+BLUEAPI_CACHE_LOCATION = "~/.cache/blueapi_cache"
+
+
+class CacheManager(ABC):
+    @abstractmethod
+    def save_cache(self, cache: Cache) -> None: ...
+    @abstractmethod
+    def load_cache(cache) -> Cache | None: ...
+    @abstractmethod
+    def delete_cache(self) -> None: ...
+
+
+class SessionCacheManager(CacheManager):
+    def __init__(self, token_path: Path | None) -> None:
+        self._token_path: Path = token_path if token_path else self._get_xdg_cache_dir()
+
+    @cached_property
+    def _file_path(self) -> str:
+        return os.path.expanduser(self._token_path)
+
+    def save_cache(self, cache: Cache) -> None:
+        cache_json: str = cache.model_dump_json()
+        cache_base64: bytes = base64.b64encode(cache_json.encode("utf-8"))
+        self.delete_cache()
+        with open(self._file_path, "xb") as token_file:
+            token_file.write(cache_base64)
+        os.chmod(self._file_path, 0o600)
+
+    def load_cache(self) -> Cache | None:
+        try:
+            with open(self._file_path, "rb") as cache_file:
+                cache_base64: bytes = cache_file.read()
+                cache_json: str = base64.b64decode(cache_base64).decode("utf-8")
+                return TypeAdapter(Cache).validate_json(cache_json)
+        except Exception:
+            return None
+
+    def delete_cache(self) -> None:
+        Path(self._file_path).unlink(missing_ok=True)
+
+    def _get_xdg_cache_dir(self) -> Path:
+        """
+        Return the XDG cache directory.
+        """
+        cache_dir = os.environ.get("XDG_CACHE_HOME")
+        if not cache_dir:
+            cache_dir = os.path.expanduser(BLUEAPI_CACHE_LOCATION)
+            if cache_dir.startswith("~/"):  # Expansion failed.
```

Comment:
This test checks whether the `~/` relative expansion has failed. If the path is absolute, it will remain unchanged. Using relative paths is simply more convenient than using absolute ones.